packages/gatsby/src/redux/persist.ts

```diff
@@ -1,12 +1,77 @@
 import v8 from "v8"
-import { readFileSync, writeFileSync } from "fs-extra"
-import { ICachedReduxState } from "./types"
+import { readFileSync, writeFileSync, unlinkSync } from "fs-extra"
+import { IReduxNode, ICachedReduxState } from "./types"
+import { sync as globSync } from "glob"
 
-const file = (): string => `${process.cwd()}/.cache/redux.state`
+const CWD = process.cwd()
+const file = (): string => CWD + `/.cache/redux.state`
+const chunkFilePrefix = (): string => CWD + `/.cache/redux.node.state_`
 
-export const readFromCache = (): ICachedReduxState =>
-  v8.deserialize(readFileSync(file()))
+export const readFromCache = (): ICachedReduxState => {
+  // The cache is stored in two steps; the nodes in chunks and the rest
+  // First we revive the rest, then we inject the nodes into that obj (if any)
+  // Each chunk is stored in its own file, this circumvents max buffer lengths
+  // for sites with a _lot_ of content. Since all nodes go into a Map, the order
+  // of reading them is not relevant.
+
+  const obj: ICachedReduxState = v8.deserialize(readFileSync(file()))
+
+  // Note: at 1M pages, this will be 1M/chunkSize chunks (ie. 1m/10k=100)
+  const chunks = globSync(chunkFilePrefix() + `*`).map(file =>
+    v8.deserialize(readFileSync(file))
+  )
+
+  const nodes: [string, IReduxNode][] = [].concat(...chunks)
```

Comment:
i think i'm confusing myself. But isn't this just creating a copy of the chunks array? If so, it doesn't look like it's really being used in a way that requires a copy. I'm probably missing something here.

Reply:
Yeah, this is shallow flatten. (Intentially one level) This is not just shallow copying an array :)

This was the best performant way I could think of to concat these potentially huge arrays. (A 1M page site that chunks per 10k nodes will have 100 elements here with each 10k elements).

In theory we could pop one element and use that as the left array of the concat, but I don't think that's a real improvement on any level.